### ISSUE TITLE
CPU Feature Detection

### ISSUE THREAD

**Issue by jncraton** [marella/ctransformers#64]
The pre-compiled binary used for inference is currently selected either manually (by setting `lib=avx` or similar) or auto selected based on the current platform:

https://github.com/marella/ctransformers/blob/41abee5aaac84ae9885f2fd90f70dcd8b3edfdfb/ctransformers/lib.py#L14-L24

This auto-selection is unfortunately too coarse to work perfectly, which results in `Illegal Instruction` errors and requires users on certain systems to manually select a binary based on their CPU feature set.

Would there be interest in a PR to support more fine-grained automatic CPU feature detection? Something like [cpuinfo](https://github.com/workhorsy/py-cpuinfo) could be used to parse CPU flags from supported operating systems. From there, the implementation logic should be straightforward:

```python
flags = cpuinfo.get_cpu_info['flags']

if 'avx2' in flags:
    path = 'avx2'
elif 'avx' in flags and 'f16c' in flags:
    path = 'avx'
else:
    path = 'basic'
```

**Comment by marella** [comment]
This looks good. In the past, I thought of using a similar library (forgot the name) but didn't use it as it is based on C and requires compiling on user's system. Since the library you mentioned doesn't require compilation, we can use it.

The logic you mentioned can be added inside the last `else` condition to avoid unnecessarily loading/calling the library when not needed.

```py
else:
    from cpuinfo import get_cpu_info
    flags = get_cpu_info()["flags"]
    if "avx2" in flags:
        path = "avx2"
    elif "avx" in flags and "f16c" in flags:
        path = "avx"
    else:
        path = "basic"
```

It would be great if you can send a PR.

**Comment by jncraton** [comment]
Sounds good. I've opened a PR (#66) for this.

**Comment by marella** [comment]
Thanks. This is released in the latest version 0.2.16

**Comment by metal3d** [comment]
Not sure if it's related but on my side, I had to recompile the package to allow AVX (only) instructions. My CPU is a bit old, even using good GPU. The "cuda" package option yields the same error. I had to recompile it myself.

Is the test, made in this PR, working?
